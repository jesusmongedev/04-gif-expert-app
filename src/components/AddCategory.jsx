import React from 'react'

function AddCategory ({ setCategory }) {
  const handleSubmit = (e) => {
    const { value } = e.target.category
    setCategory(value)
    e.preventDefault()
  }

  return (
    <form onSubmit={handleSubmit}>
      <div className='flex'>
        <input className='flex-1 px-4' type='search' name='category' />
        <button className='bg-gray-300 text-slate-700 px-3 py-2'>
          Buscar
        </button>
      </div>
    </form>
  )
}

export default AddCategory
