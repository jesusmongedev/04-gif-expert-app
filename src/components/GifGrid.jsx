import React from 'react'
import Proptypes from 'prop-types'

import { GifGridItem, GifGridItemSkeleton } from './'
import useGetGifs from './../hooks/useGetGifs'

export function GifGrid ({ handleDelete, category }) {
  const { images, error, isLoading } = useGetGifs({ category })

  return (
    <div>
      <div className='flex justify-between items-center space-x-4 md:space-x-8'>
        <h3 className='text-xl md:text-2xl font-bold mb-6 mt-10 flex-1'>{category}</h3>
        <button onClick={() => handleDelete(category)}>üóëÔ∏è</button>
      </div>
      {error && (
        <div className='text-center'>
          <p className='text-red-500 font-bold'>Error</p>
          <p>{error.message}</p>
        </div>
      )}
      <div className='sm:columns-2 lg:columns-3 gap-6 lg:gap-8'>
        {isLoading && [1, 2, 3].map((n) => <GifGridItemSkeleton key={n} />)}
        {images.map(({ id, title, url }) => (
          <GifGridItem
            key={id}
            url={url}
            title={title}
          />
        ))}
      </div>
    </div>
  )
}

GifGrid.propTypes = {
  handleDelete: Proptypes.func.isRequired,
  category: Proptypes.string.isRequired
}
