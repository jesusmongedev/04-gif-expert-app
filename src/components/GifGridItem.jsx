import React from 'react'
import Proptypes from 'prop-types'

export function GifGridItem ({ url, title }) {
  return (
    <div className='w-full aspect-video mb-6 lg:mb-10 bg-gray-100 rounded  shadow-lg'>
      <img className='w-full' src={url} alt={title} />
      <div className='px-6 py-4'>
        <div className='font-bold text-xl mb-2'>{title}</div>
      </div>
    </div>
  )
}

GifGridItem.propTypes = {
  url: Proptypes.string.isRequired,
  title: Proptypes.string.isRequired
}
