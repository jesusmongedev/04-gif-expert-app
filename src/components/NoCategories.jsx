import React from 'react'

export function NoCategories () {
  return (
    <div className='mt-40 text-center'>
      <p className='text-red-500 font-bold'>No categories</p>
      <p>Add a new category to start</p>
    </div>
  )
}
