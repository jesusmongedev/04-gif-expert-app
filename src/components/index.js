export * from './AddCategory'
export * from './GifGrid'
export * from './GifGridItem'
export * from './NoCategories'
export * from './Layout/GifGridItemSkeleton'
