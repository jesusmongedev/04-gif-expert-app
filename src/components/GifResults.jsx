import useGetGifs from '@/hooks/useGetGifs'
import React from 'react'
import GifGrid from './GifGrid'

function GifResults ({ category }) {
  const { images, error, loading, setImages } = useGetGifs({ category })
  console.log('images', images)

  const handleDelete = (category) => {
    const { [category]: deleted, ...rest } = images
    setImages(rest)
  }

  if (loading) {
    return (
      <div className='flex justify-center items-center mt-56'>
        <div className='animate-spin rounded-full h-32 w-32 border-t-2 border-b-2 border-gray-900' />
      </div>
    )
  }

  if (error) {
    return (
      <div className='text-center mt-56'>
        <p className='text-red-500 font-bold'>Error</p>
        <p>{error.message}</p>
      </div>
    )
  }

  return (
    <>
      <div>
        {Object.keys(images)
          .reverse()
          .map((category) => (
            <GifGrid
              key={category}
              handleDelete={handleDelete}
              category={category}
              images={images}
            />
          ))}
      </div>
    </>
  )
}

export default GifResults
