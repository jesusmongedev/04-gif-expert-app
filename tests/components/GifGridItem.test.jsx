import React from 'react'
import { GifGridItem } from './../../src/components/GifGridItem'
import { render, screen } from '@testing-library/react'
describe('Tests for GifGridItem component', () => {
  // Arrange
  const title = 'Brian Tracy'
  const url = 'https://www.briantracy.com/blog/wp-content/uploads/2019/07/what-is-success.jpg'
  // Act
  let _container
  beforeEach(() => {
    const { container } = render(<GifGridItem title={title} url={url} />)
    _container = container
  })
  // screen.debug()
  // Assert
  test('component should Match Snapshot', () => {
    expect(_container).toMatchSnapshot()
  })

  test('should show the img with indicated props', () => {
    const { src, alt } = screen.getByRole('img')
    expect(src).toBe(url)
    expect(alt).toBe(title)
  })

  test('should render the title', () => {
    const hasTitle = screen.getByText(title)
    expect(hasTitle).toBeTruthy()
  })
})
