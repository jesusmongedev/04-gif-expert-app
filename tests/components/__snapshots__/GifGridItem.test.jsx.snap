// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for GifGridItem component component should Match Snapshot 1`] = `
<div>
  <div
    class="w-full aspect-video mb-6 lg:mb-10 bg-gray-100 rounded  shadow-lg"
  >
    <img
      alt="Brian Tracy"
      class="w-full"
      src="https://www.briantracy.com/blog/wp-content/uploads/2019/07/what-is-success.jpg"
    />
    <div
      class="px-6 py-4"
    >
      <div
        class="font-bold text-xl mb-2"
      >
        Brian Tracy
      </div>
    </div>
  </div>
</div>
`;
